### PR TITLE
Keyboard Support of the Razer Blade Pro Early 2020

### PR DESCRIPTION
**Pull request by Baiyah7o**
Hello,

this adds support for the keyboard of the Razer Blade Pro Early 2020.
```
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0256 RZ09--0329, Gaming Laptop [Blade Pro 17 Full HD (2020)]
```

As in the request _Support for Blade Stealth Late 2020 #1371_ I have started using occurrences for the Stealth Early 2020 for the Blade Pro Early 2020 (1532:0256). Most worked as expected when tested in Razercommander, Polychromatic, RazerGenie and with the scripts from the examples folder.

- **Spectrum, Wave, Reactive, Ripple, Static and Custom work**. _(Static sometimes requires a second submit.)_
- Breathing does not work at all 
- Starlight only works in **single green colour**.
- Custom colour settings do not work in Razercommander

Best wishes,
Patrick

### CODE REVIEW

**Comment by z3ntu** [comment]
I'll keep this open until the open points are resolved, diff looks sane otherwise.

**Comment by z3ntu** [comment]
What's the status here?

**Comment by Baiyah7o** [comment]
Updated the fork to v3.0.0, It looks fine, but  'Mark resolved' is grayed out for me.

**Comment by z3ntu** [comment]
Can you send an updated list for what you wrote in the initial message?

> * Spectrum, Wave, Reactive, Ripple, Static and Custom work. (Static sometimes requires a second submit.)
> * Breathing does not work at all
> * Starlight only works in single green colour.
> * Custom colour settings do not work in Razercommander

Also if it's too much work for you to resolve the conflicts, I can do that while merging.

**Comment by Baiyah7o** [comment]
No chages to the initial message
- Spectrum, Wave, Reactive, Ripple, Static and Custom work. (Static still requires a second submit.)
- Breathing does not work at all
- Starlight only works in single green colour.
- Custom colour settings do not work in Razercommander

File: README.md [hunk 1]
```diff
@@ -67,6 +67,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Blade Stealth (Early 2020)              |  1532:0252  |
 | Razer Blade 15 Advanced (2020)                |  1532:0253  |
 | Razer Blade 15 (Early 2020) Base              |  1532:0255  |
+| Razer Blade Pro (Early 2020)                  |  1532:0256  |
```
Comment (decryp2kanon) [inline README.md]: i guess this model has 17 inch screen
```suggestion
| Razer Blade Pro 17 (Early 2020)                  |  1532:0256  |
```

**Comment by decryp2kanon** [comment]
unfortunately not working on mine.

> razer blade pro 17 2020
Bus 001 Device 003: ID 1532:0256 Razer USA, Ltd Razer Blade
Ubuntu 20.04.2 LTS
Linux neiru 5.8.0-45-generic
Polychromatic v0.3.12-stable

what i tried
```bash
git clone https://github.com/openrazer/openrazer.git
cd openrazer
git fetch origin pull/1376/head:test_1
git checkout test_1
./scripts/build_debs.sh
sudo apt install ./dist/*.deb
# REBOOT
```

image

how can i test this PR?

**Comment by Baiyah7o** [comment]
> Arch Linux (kernels 5.11.9.arch1-1, lts 5.10.26-1 and zen 5.11.9.zen1-1)
> Product Number: RZ09-03295E42
> Model Number:   RZ09-0329
> Bus 001 Device 003: ID 1532:0256 Razer USA, Ltd Razer Blade
> polychromatic v0.6.0

Did `udev/99-razer.rules` update properly? I remember having similar issues _(Device listed but not recognized.)_ when the `MODULE_ALIAS("hid:b0003g*v00001532p00000256");
` line was missing in `driver/razerkbd.mod.c`. 

_I am on v0.6.0 of polychromatic since v0.3.12 gui is completely broken by Python upgrades earlier on my system. But the device is listed and hitting on some of the remaining  buttons in the gui indicates that it works beside the gui._ 

I did a clean install again and can not reproduce the error. So far I always installed manually, and i never built debs.

```
git clone https://github.com/openrazer/openrazer.git
cd openrazer
git fetch origin pull/1376/head:test_1
git checkout test_1

make appstream_install
make daemon_install
make driver
make driver_install
make install-systemd
make manual_install_msg
make python_library_install
make setup_dkms
make udev_install
make xdg_install
dkms install openrazer-driver -v 3.0.0
```
### Polychromaticv0.6.0
polychromatic_controller
polychromatic_deviceinfo
polycromatic_version
### Broken Polychromatic v0.3.12
polychromatic-0 3 12
### Razercommander
razercommander
### RazerGenie
razergenie
### Snake
snake

**Comment by decryp2kanon** [comment]
> Did udev/99-razer.rules update properly? I remember having similar issues (Device listed but not recognized.) when the MODULE_ALIAS("hid:b0003g*v00001532p00000256"); line was missing in driver/razerkbd.mod.c.

how to update udev? i am a noob :sweat_smile: 

can you give me a step by step method?

**Comment by decryp2kanon** [comment]
> make manual_install_msg

```
$ sudo make manual_install_msg 
Please do not install the driver using this method. Use a distribution package as it tracks the files installed and can remove them afterwards. If you are 100% sure, you want to do this, find the correct target in the Makefile.
Exiting.
```

can i ignore this message?

**Comment by decryp2kanon** [comment]
something went wrong, and my keyboard doesn't work more. i am reinstall ubuntu 20.04 lol

**Comment by decryp2kanon** [comment]
i tested my own, but still keyboard not working after load openrazer. i have no idea...

https://github.com/openrazer/openrazer/compare/master...decryp2kanon:master

**Comment by Baiyah7o** [comment]
It installed successfully on Ubuntu 20.04 LTS and Ubuntu 18.04 LTS.  dpkg required some additional packages during install to build modules. 

`sudo apt install python3-pyudev python3-setproctitle python3-notify2 python3-daemonize xautomation
`

Otherwise used same install command as you did before. 

```
git clone https://github.com/openrazer/openrazer.git
cd openrazer
git fetch origin pull/1376/head:test_1
git checkout test_1
./scripts/build_debs.sh
sudo apt install ./dist/*.deb
```
Polychromatic v0.3.12 main windows is still looks broken but the tray apps is fully functional.

polychromatic-lts
polychromatic-lts-tray
_(Screenshots taken from LTS 18.04)_

**Comment by lah7** [comment]
> Polychromatic v0.3.12 main windows is still looks broken but the tray apps is fully functional.

The earlier _broken_ screenshot on Arch looks like v0.3.12 just needed a reinstall (styles missing).

 v0.3.12's Controller has a bug with newer Blade laptops (it can be patched via https://github.com/polychromatic/polychromatic/issues/229). v0.6.0 would be a better version for testing support.

:bulb: v0.6.0 has new keyboard graphics for Razer Blade (2016 & 2018 models) for the effect editor. Hopefully one of them matches up with the 2020 models too.

---

By the way, it looks like `build` directories were accidentally committed in this pull request 🗑️

**Comment by decryp2kanon** [comment]
> v0.6.0 would be a better version for testing support.

```
sudo add-apt-repository ppa:polychromatic/edge
sudo apt install polychromatic
```

i tried this v0.6.0 PPA. and ubuntu totally broken lol

how to install v0.6.0?

**Comment by lah7** [comment]
@decryp2kanon That should be it :+1: Ubuntu shouldn't "break" because of one app, though this (and any other frontend) might not work if OpenRazer is broken.  Using the `makefile` directly for example is **not recommended** because Apt can't keep track of what was installed, and there could be some leftovers interfering.

Similar to @Baiyah7o instructions, testing this branch on Ubuntu can be done like this:
```
git clone https://github.com/Baiyah7o/openrazer.git -b master
cd openrazer
sudo apt install dh-python dkms python3-setuptools
./scripts/build_debs.sh
sudo apt install ./dist/*.deb
sudo gpasswd -a $USER plugdev
# Reboot
```

Maybe everything is installed for you but can't be loaded because of secure boot (which must be turned off in BIOS/EFI settings). Have a look at the [troubleshooting guide](https://github.com/openrazer/openrazer/wiki/Troubleshooting) if you're still having issues.

**Comment by decryp2kanon** [comment]
> git clone https://github.com/Baiyah7o/openrazer.git -b master
cd openrazer
sudo apt install dh-python dkms python3-setuptools
./scripts/build_debs.sh
sudo apt install ./dist/*.deb
sudo gpasswd -a $USER plugdev
Reboot

i tried on vbox, but i got after reboot
image

after force reboot again,
image

anyway it seems working. thanks!

**Comment by lah7** [comment]
> FATAL: Keyboard error:995

Glad it works! Looks like that error is [a bug or race condition in VirtualBox](https://reactos.org/archives/public/ros-bugs/2017-November/136438.html) and no fault of the driver.

**Comment by jpgls** [comment]
> something went wrong, and my keyboard doesn't work more. i am reinstall ubuntu 20.04 lol

Obviously you've resolved this by now, but I ran into this same issue earlier today while working on a version for my machine - if you run into this again, I was able to turn on the on screen keyboard, then login and `apt remove openrazer-meta openrazer-daemon openrazer-doc openrazer-drier-dkms` - a reboot after that brought my keyboard back to life without needing to reinstall my OS.

**Review by Baiyah7o — COMMENTED** [review]
Looks good to me, builds and runs on own laptop.

**Comment by z3ntu** [comment]
> No chages to the initial message
> 
> * Spectrum, Wave, Reactive, Ripple, Static and Custom work. (Static still requires a second submit.)
> * Breathing does not work at all
> * Starlight only works in single green colour.
> * Custom colour settings do not work in Razercommander

Still the case? The static one needs to be fixed, if breathing/starlight won't work then remove the effects, I won't merge broken features.

**Comment by Baiyah7o** [comment]
I also removed the logo because the logo on my device is broken and I am unable to verify functionality.

**Review by z3ntu — APPROVED** [review]
I've fixed up a few remaining things, LGTM now, thanks!

**Pull request merged**